### PR TITLE
Add Boundary Attack++

### PR DESCRIPTION
**Pull request by Jianbo-Lab**
Add a decision-based attack, Boundary Attack++, from the preprint: https://arxiv.org/abs/1904.02144.

### CODE REVIEW

File: cleverhans/attacks/bapp.py [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+    self.gamma = gamma
+    self.constraint = constraint
+    self.batch_size = batch_size
+    self.clip_min = clip_min
+    self.clip_max = clip_max
+    self.verbose = verbose
+
+
+def _check_first_dimension(x, tensor_name):
+  message = "Require {} has batch_size of 1.".format(tensor_name)
```
Comment (npapernot) [inline cleverhans/attacks/bapp.py]:
```suggestion
  message = "Tensor {} should have batch_size of 1.".format(tensor_name)
```

File: cleverhans/attacks/bapp.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+  if x.get_shape().as_list()[0] is None:
+    check_batch = utils_tf.assert_equal(tf.shape(x)[0], 1, message=message)
+    with tf.control_dependencies([check_batch]):
+      x = tf.identity(x)
+  elif x.get_shape().as_list()[0] != 1:
+    raise ValueError(message)
+
+
+class BAPP(object):
```
Comment (npapernot) [inline cleverhans/attacks/bapp.py]: What is the reason for having two separate classes, with the first class basically (if I understand correctly) wrapping the second class?

File: cleverhans/attacks/bapp.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+      dist = compute_distance(perturbed, sample, self.constraint)
+      if self.verbose:
+        print('iteration: {:d}, {:s} distance {:.4E}'.format(
+            j+1, self.constraint, dist))
+
+    perturbed = np.expand_dims(perturbed, 0)
+    return perturbed
+
+
+def clip_image(image, clip_min, clip_max):
```
Comment (npapernot) [inline cleverhans/attacks/bapp.py]: You can probably reuse: https://github.com/tensorflow/cleverhans/blob/f70ca7e000dadd6ace5aeff15bba0e960e8c1384/cleverhans/utils_tf.py#L540

File: cleverhans/attacks/bapp.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+    fval = fval - np.mean(fval)
+    gradf = np.mean(fval * rv, axis=0)
+
+  # Get the gradient direction.
+  gradf = gradf / np.linalg.norm(gradf)
+
+  return gradf
+
+
+def project(original_image, perturbed_images, alphas, shape, constraint):
```
Comment (npapernot) [inline cleverhans/attacks/bapp.py]: You can probably reuse https://github.com/tensorflow/cleverhans/blob/f70ca7e000dadd6ace5aeff15bba0e960e8c1384/cleverhans/utils_tf.py#L358

File: cleverhans/attacks/bapp.py [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+        perturbed_images,
+        original_image - alphas,
+        original_image + alphas
+    )
+  return projected
+
+
+def binary_search_batch(original_image, perturbed_images, decision_function,
+                        shape, constraint, theta):
+  """ Binary search to approach the boundar. """
```
Comment (npapernot) [inline cleverhans/attacks/bapp.py]:
```suggestion
  """ Binary search to approach the boundary. """
```

File: tests_tf/test_attacks.py [hunk 6]
```diff
@@ -1281,3 +1283,137 @@ def test_keep_pixels_on_image(self):
...
+    self.assertTrue(np.mean(orig_labs == new_labs) < 0.1)
+
+  def test_generate_untargeted_linf(self):
+
+    x_val = np.random.rand(50, 2)
+    x_val = np.array(x_val, dtype=np.float32)
+
+    orig_labs = np.argmax(self.sess.run(self.model.get_logits(x_val)), axis=1)
+
+    # Require input has batchsize 1.
```
Comment (npapernot) [inline tests_tf/test_attacks.py]:
```suggestion
    # Requires input to have batchsize 1.
```

File: cleverhans/attacks/bapp.py (hunk 3 above)
Comment (Jianbo-Lab) [reply cleverhans/attacks/bapp.py]: clip_by_value only applies to clipping by 0-dim tensor. clip_image is capable of clipping by a tensor with the same shape as the image. Also, as the wrapper is used, it might be more consistent to not use tf operators in the function.

File: cleverhans/attacks/bapp.py (hunk 4 above)
Comment (Jianbo-Lab) [reply cleverhans/attacks/bapp.py]: 'clip_eta' is not applicable here. 'clip_eta' with ord=2 first calculates the norm of a tensor, and then divides the tensor by the norm. Thus clip_eta is equivalent to projection onto the unit sphere centered at the origin. 'project' projects a tensor to a sphere centered at a given point (not necessarily origin) with a given radius (not necessarily its norm). 'project' seems to be a more general version of 'clip_eta'.

File: cleverhans/attacks/bapp.py (hunk 2 above)
Comment (Jianbo-Lab) [reply cleverhans/attacks/bapp.py]: I tried to follow the CleverHans code of C&W attack when a wrapper is used: https://github.com/tensorflow/cleverhans/blob/master/cleverhans/attacks/carlini_wagner_l2.py
But I agree with your suggestion that the extra class is redundant. I've removed it in the updated version. Thanks Nicolas!

**Comment by npapernot** [comment]
Thanks for the PR! Feel free to add your name to the author list in the README @Jianbo-Lab

**Comment by Jianbo-Lab** [comment]
Cool! I've created another PR for updating README.md. @npapernot
